### ISSUE TITLE
Feature Request: publish marionette protocol to crates.io

### ISSUE THREAD

**Issue by brycefisher** [mozilla/geckodriver#1382]
I'm trying to build a rust project which only targets Firefox. I know that mozilla central holds geckodriver source code, but that only the webdriver facade (webdriver crate) is published to crates.io. I basically just want the other half of geckodriver -- I want to have a library which speaks the Firefox remote protocol but doesn't know about webdriver.

Breaking changes / strong backwards compatibility with old Firefox is not important to me. I'm happy to chase.

Afaict, my own option right would be to manually copy out the relevant marionette protocol bits and bobs into my project.

Is it crazy to dream that a marionette crate could be extracted and published to crates.io?

**Comment by whimboo** [comment]
I think that something we wanted to do anyway, and I even did some preparation work for that lately. So I filed https://bugzilla.mozilla.org/show_bug.cgi?id=1493095 for the remaining work.

**Comment by andreastt** [comment]
Yes! We have the intention to also separate out the Marionette client to a separate crate. As you know, we recently [moved from rustc_serialize to serde](https://bugzilla.mozilla.org/show_bug.cgi?id=1396821) which was a major blocker for making something like this happen.

**Comment by brycefisher** [comment]
Amazing! Can I contribute to this effort in any way as a non Mozilla employee?

**Comment by whimboo** [comment]
Sure, and that is great to hear! If you have the time and willingness just hop over to Bugzilla, create an account, and mention your interest to work on this bug. We can then mentor you through the whole process.

**Comment by andreastt** [comment]
I started looking at this over the weekend for a bit of fun, and it turns out that this is not a particularly hard bug to solve but it requires some in-depth knowledge of the Marionette protocol. 

We do have some developer documentation here:
https://firefox-source-docs.mozilla.org/testing/geckodriver/geckodriver/index.html#for-developers

**Comment by whimboo** [comment]
Please lets move any further discussion to Bugzilla. Thanks.
